### PR TITLE
support vpc configuration of multiple external network segments through label and crd

### PR DESCRIPTION
**Pull request by zcq98**
- [x] Make sure you have followed [Kube-OVN Code Style](https://github.com/kubeovn/kube-ovn/blob/master/CODE_STYLE.md).

### What type of this PR
Examples of user facing changes:
- Features

### Which issue(s) this PR fixes:
Fixes #(issue-number)

### <samp>🤖 Generated by Copilot at 6ac6836</samp>

This pull request adds support for multiple external subnets for a VPC in the `kubeovn` API and controller. It modifies the `VpcSpec` and `VpcStatus` structs, the `vpc.go` and `subnet.go` files, and comments out the BFD and custom VPC static route features, which are incompatible with the new feature.

### <samp>🤖 Generated by Copilot at 6ac6836</samp>

> _No more static routes, no more BFD_
> _We unleash the power of multiple subnets_
> _We reconcile the VPCs with a new spec_
> _We handle the gateways with a new logic_

### <samp>🤖 Generated by Copilot at 6ac6836</samp>

*  Add support for multiple external subnets for a VPC ([link](https://github.com/kubeovn/kube-ovn/pull/3264/files?diff=unified&w=0#diff-69f75d53361877cc02ab6f413f306486d29287c6400b92b71c4c74f018f3234aL418-R424), [link](https://github.com/kubeovn/kube-ovn/pull/3264/files?diff=unified&w=0#diff-69f75d53361877cc02ab6f413f306486d29287c6400b92b71c4c74f018f3234aR485), [link](https://github.com/kubeovn/kube-ovn/pull/3264/files?diff=unified&w=0#diff-ae8a8b5d44d277a4bb96635ee325fe54b48e8bc10b24a79d4bb8e211b57131ecR9), [link](https://github.com/kubeovn/kube-ovn/pull/3264/files?diff=unified&w=0#diff-ae8a8b5d44d277a4bb96635ee325fe54b48e8bc10b24a79d4bb8e211b57131ecL118-R132), [link](https://github.com/kubeovn/kube-ovn/pull/3264/files?diff=unified&w=0#diff-ae8a8b5d44d277a4bb96635ee325fe54b48e8bc10b24a79d4bb8e211b57131ecL529-R652), [link](https://github.com/kubeovn/kube-ovn/pull/3264/files?diff=unified&w=0#diff-ae8a8b5d44d277a4bb96635ee325fe54b48e8bc10b24a79d4bb8e211b57131ecL1003-R1092), [link](https://github.com/kubeovn/kube-ovn/pull/3264/files?diff=unified&w=0#diff-ae8a8b5d44d277a4bb96635ee325fe54b48e8bc10b24a79d4bb8e211b57131ecL1021-R1108), [link](https://github.com/kubeovn/kube-ovn/pull/3264/files?diff=unified&w=0#diff-ae8a8b5d44d277a4bb96635ee325fe54b48e8bc10b24a79d4bb8e211b57131ecL1039-R1165), [link](https://github.com/kubeovn/kube-ovn/pull/3264/files?diff=unified&w=0#diff-ae8a8b5d44d277a4bb96635ee325fe54b48e8bc10b24a79d4bb8e211b57131ecR1180-R1190), [link](https://github.com/kubeovn/kube-ovn/pull/3264/files?diff=unified&w=0#diff-ae8a8b5d44d277a4bb96635ee325fe54b48e8bc10b24a79d4bb8e211b57131ecL1128-R1244), [link](https://github.com/kubeovn/kube-ovn/pull/3264/files?diff=unified&w=0#diff-ae8a8b5d44d277a4bb96635ee325fe54b48e8bc10b24a79d4bb8e211b57131ecR1271-R1283))
  * Add a new field `ExternalSubnets` to the `VpcSpec` and `VpcStatus` structs in `types.go` to store the desired and actual external subnets for a VPC ([link](https://github.com/kubeovn/kube-ovn/pull/3264/files?diff=unified&w=0#diff-69f75d53361877cc02ab6f413f306486d29287c6400b92b71c4c74f018f3234aL418-R424), [link](https://github.com/kubeovn/kube-ovn/pull/3264/files?diff=unified&w=0#diff-69f75d53361877cc02ab6f413f306486d29287c6400b92b71c4c74f018f3234aR485))
  * Import the `sort` package in `vpc.go` to sort the external subnets in the `handleAddOrUpdateVpc` function ([link](https://github.com/kubeovn/kube-ovn/pull/3264/files?diff=unified&w=0#diff-ae8a8b5d44d277a4bb96635ee325fe54b48e8bc10b24a79d4bb8e211b57131ecR9))
  * Modify the `handleDelVpc` function in `vpc.go` to loop over the external subnets in the VPC spec and call `handleDelVpcExternalSubnets` for each subnet ([link](https://github.com/kubeovn/kube-ovn/pull/3264/files?diff=unified&w=0#diff-ae8a8b5d44d277a4bb96635ee325fe54b48e8bc10b24a79d4bb8e211b57131ecL118-R132))
  * Modify the `handleAddOrUpdateVpc` function in `vpc.go` to check the external subnets in the VPC spec and status, and call `handleAddVpcExternalSubnets` or `handleDelVpcExternalSubnets` for each subnet that is added or removed ([link](https://github.com/kubeovn/kube-ovn/pull/3264/files?diff=unified&w=0#diff-ae8a8b5d44d277a4bb96635ee325fe54b48e8bc10b24a79d4bb8e211b57131ecL529-R652))
  * Rename the `handleAddVpcExternal` function in `vpc.go` to `handleAddVpcExternalSubnets` and add a parameter `subnet` to specify the external subnet to connect ([link](https://github.com/kubeovn/kube-ovn/pull/3264/files?diff=unified&w=0#diff-ae8a8b5d44d277a4bb96635ee325fe54b48e8bc10b24a79d4bb8e211b57131ecL1003-R1092))
  * Change the references to the `c.config.ExternalGatewaySwitch` to the `subnet` parameter in the `handleAddVpcExternalSubnets` function in `vpc.go`, and replace the logic to get the gateway chassis from the config map to the node labels ([link](https://github.com/kubeovn/kube-ovn/pull/3264/files?diff=unified&w=0#diff-ae8a8b5d44d277a4bb96635ee325fe54b48e8bc10b24a79d4bb8e211b57131ecL1003-R1092))
  * Change the call to `c.acquireIPAddress` and `c.createOrUpdateCrdOvnEip` in the `handleAddVpcExternalSubnets` function in `vpc.go` to use the `subnet` parameter instead of the `c.config.ExternalGatewaySwitch` ([link](https://github.com/kubeovn/kube-ovn/pull/3264/files?diff=unified&w=0#diff-ae8a8b5d44d277a4bb96635ee325fe54b48e8bc10b24a79d4bb8e211b57131ecL1021-R1108))
  * Change the call to `c.OVNNbClient.CreateLogicalPatchPort` in the `handleAddVpcExternalSubnets` function in `vpc.go` to use the `subnet` parameter instead of the `c.config.ExternalGatewaySwitch`, and the `lspName` and `lrpName` variables to use the `subnet` parameter instead of the `c.config.ExternalGatewaySwitch` ([link](https://github.com/kubeovn/kube-ovn/pull/3264/files?diff=unified&w=0#diff-ae8a8b5d44d277a4bb96635ee325fe54b48e8bc10b24a79d4bb8e211b57131ecL1039-R1165))
  * Append the external subnet to the `vpc.Status.ExternalSubnets` slice in the `handleAddVpcExternalSubnets` function in `vpc.go` if it is not already in the slice ([link](https://github.com/kubeovn/kube-ovn/pull/3264/files?diff=unified&w=0#diff-ae8a8b5d44d277a4bb96635ee325fe54b48e8bc10b24a79d4bb8e211b57131ecR1180-R1190))
  * Rename the `handleDelVpcExternal` function in `vpc.go` to `handleDelVpcExternalSubnets` and add a parameter `subnet` to specify the external subnet to disconnect ([link](https://github.com/kubeovn/kube-ovn/pull/3264/files?diff=unified&w=0#diff-ae8a8b5d44d277a4bb96635ee325fe54b48e8bc10b24a79d4bb8e211b57131ecL1128-R1244))
  * Change the references to the `c.config.ExternalGatewaySwitch` to the `subnet` parameter in the `handleDelVpcExternalSubnets` function in `vpc.go`, and the `lspName` and `lrpName` variables to use the `subnet` parameter instead of the `c.config.ExternalGatewaySwitch` ([link](https://github.com/kubeovn/kube-ovn/pull/3264/files?diff=unified&w=0#diff-ae8a8b5d44d277a4bb96635ee325fe54b48e8bc10b24a79d4bb8e211b57131ecL1128-R1244))
  * Remove the external subnet from the `vpc.Status.ExternalSubnets` slice in the `handleDelVpcExternalSubnets` function in `vpc.go` if it is in the slice ([link](https://github.com/kubeovn/kube-ovn/pull/3264/files?diff=unified&w=0#diff-ae8a8b5d44d277a4bb96635ee325fe54b48e8bc10b24a79d4bb8e211b57131ecR1271-R1283))
* Comment out the call to `reconcileCustomVpcStaticRoute` in the `reconcileSubnet` function in `subnet.go`, which is used to reconcile the static routes for a custom VPC ([link](https://github.com/kubeovn/kube-ovn/pull/3264/files?diff=unified&w=0#diff-3742843fd11c637b26eae183fb8a0c4eeff96bbfb18a8db1e0ded0c33733f205L1039-R1044))
* Comment out the code related to the BFD feature in the `handleAddOrUpdateVpc` function in `vpc.go`, which is not compatible with the multiple external subnets feature ([link](https://github.com/kubeovn/kube-ovn/pull/3264/files?diff=unified&w=0#diff-ae8a8b5d44d277a4bb96635ee325fe54b48e8bc10b24a79d4bb8e211b57131ecL529-R652))

### CODE REVIEW

File: pkg/controller/subnet.go [hunk 1]
```diff
@@ -1036,12 +1036,12 @@ func (c *Controller) reconcileSubnet(subnet *kubeovnv1.Subnet) error {
 		}
 	}
 
-	if subnet.Spec.Vpc != c.config.ClusterRouter {
-		if err := c.reconcileCustomVpcStaticRoute(subnet); err != nil {
-			klog.Errorf("reconcile custom vpc ovn route for subnet %s failed: %v", subnet.Name, err)
-			return err
-		}
-	}
+	// if subnet.Spec.Vpc != c.config.ClusterRouter {
```
Comment (bobz965) [inline pkg/controller/subnet.go:1039]: 这里不需要改

File: pkg/controller/vpc.go [hunk 2]
```diff
@@ -526,47 +535,121 @@ func (c *Controller) handleAddOrUpdateVpc(key string) error {
...
+						if subnetStatus == subnetSpec {
+							found = true
+							break
+						}
+					}
+					if !found {
+						onlyInStatus = append(onlyInStatus, subnetStatus)
+					}
+				}
+				klog.Info("externalSubnets in Spec not in Status : ", onlyInSpec)
```
Comment (bobz965) [inline pkg/controller/vpc.go:590]: remove debug log

File: pkg/controller/vpc.go [hunk 3]
```diff
@@ -526,47 +535,121 @@ func (c *Controller) handleAddOrUpdateVpc(key string) error {
...
-			if err := c.OVNNbClient.DeleteBFD(lrpEipName, ""); err != nil {
-				klog.Error(err)
-				return err
-			}
-			if err := c.handleDeleteVpcStaticRoute(key); err != nil {
-				klog.Errorf("failed to delete bfd route for vpc %s, error %v", key, err)
-				return err
-			}
-		}
+			// 	if vpc.Spec.EnableBfd {
```
Comment (bobz965) [inline pkg/controller/vpc.go:611]: 这里的代码不影响，需要还原

File: pkg/controller/vpc.go [hunk 4]
```diff
@@ -526,47 +535,121 @@ func (c *Controller) handleAddOrUpdateVpc(key string) error {
...
-			if !vpc.Spec.EnableBfd {
-				// auto add normal type static route, if not use ecmp based bfd
-				klog.Infof("add normal external static route for enable external vpc %s", vpc.Name)
-				if err := c.reconcileCustomVpcAddNormalStaticRoute(vpc.Name); err != nil {
-					klog.Errorf("failed to reconcile vpc %q bfd static route", vpc.Name)
+			if cachedVpc.Status.ExternalSubnets != nil && len(cachedVpc.Status.ExternalSubnets) != 0 {
+				sort.Strings(cachedVpc.Status.ExternalSubnets)
+			}
+			if cachedVpc.Status.EnableExternal && !reflect.DeepEqual(cachedVpc.Spec.ExternalSubnets, cachedVpc.Status.ExternalSubnets) {
+				onlyInSpec := []string{}
```
Comment (bobz965) [inline pkg/controller/vpc.go:564]: 这两个数组应该可以省掉，直接基于status的当前内容判断即可

File: pkg/controller/vpc.go [hunk 5]
```diff
@@ -1154,6 +1268,19 @@ func (c *Controller) handleDelVpcExternal(key string) error {
 	}
 	vpc := cachedVpc.DeepCopy()
 	vpc.Status.EnableExternal = cachedVpc.Spec.EnableExternal
+	var index int
```
Comment (bobz965) [inline pkg/controller/vpc.go:1271]: for loop 确认status中有的subnet 而subnet 不存在的时候直接删就行，不需要记录个index 再删

**Review by bobz965 — CHANGES_REQUESTED** [review]

File: pkg/controller/vpc.go [hunk 6]
```diff
@@ -1036,30 +1119,50 @@ func (c *Controller) handleAddVpcExternal(key string) error {
...
+		klog.Errorf("failed to list nodes, %v", err)
 		return err
 	}
-	if cm.Data["enable-external-gw"] == "false" {
-		err := fmt.Errorf("gw chassis included in config map %s not exist", util.ExternalGatewayConfig)
-		// TODO://
-		// gw chassis controlled by a crd is a better way
-		// gw chassis crd may generated by this cm, or by user manually, or by node ext gw type oeip
+	if nodes == nil {
+		err := fmt.Errorf("failed to get chassis nodes, there is no labeled gw node")
```
Comment (bobz965) [inline pkg/controller/vpc.go:1130]: 这段应该不需要，过滤出来的已经是gw nodes了

**Review by bobz965 — CHANGES_REQUESTED** [review]

File: pkg/controller/vpc.go (hunk 3 above)
Comment (zcq98) [reply pkg/controller/vpc.go:611]: 这里如果不注释的话，只要没有配置name为external的外部网段，这边直接返回err，后续关闭EnableExternal的流程就走不到了

File: pkg/controller/vpc.go (hunk 4 above)
Comment (zcq98) [reply pkg/controller/vpc.go:564]: 这里有可能用户会删除原有的external subnet并且添加新的external subnet，比如spec：A，B -> B，C，这样的话我既要删除A，又要添加B，这个只基于status好像不行哇

File: pkg/controller/vpc.go (hunk 6 above)
Comment (zcq98) [reply pkg/controller/vpc.go:1130]: 如果没有node有ovn.kubernetes.io/external-gw的标记，c.nodesLister.List(sel)不会返回err，node此时为nil，这样的话chassis就为空了

**Comment by lelvisl** [comment]
@zcq98 this is a really cool feature! i'm waiting for merging this pr =)

File: pkg/controller/vpc.go [hunk 7]
```diff
@@ -549,6 +562,49 @@ func (c *Controller) handleAddOrUpdateVpc(key string) error {
 					return err
 				}
 			}
+			if cachedVpc.Spec.ExternalSubnets != nil {
+				sort.Strings(cachedVpc.Spec.ExternalSubnets)
+			}
+			if cachedVpc.Status.ExternalSubnets != nil {
```
Comment (bobz965) [inline pkg/controller/vpc.go:568]: 如果status是spec赋值得到的，则不需要这一次sort

File: pkg/controller/vpc.go [hunk 8]
```diff
@@ -549,6 +562,49 @@ func (c *Controller) handleAddOrUpdateVpc(key string) error {
...
+				}
+				for _, subnetStatus := range cachedVpc.Status.ExternalSubnets {
+					found := false
+					for _, subnetSpec := range cachedVpc.Spec.ExternalSubnets {
+						if subnetStatus == subnetSpec {
+							found = true
+							break
+						}
+					}
+					if !found {
```
Comment (bobz965) [inline pkg/controller/vpc.go:596]: 添加和删除应该可以放入一个循环里

File: pkg/controller/vpc.go [hunk 9]
```diff
@@ -1074,6 +1162,19 @@ func (c *Controller) handleAddVpcExternal(key string) error {
...
+	if subnet != c.config.ExternalGatewaySwitch {
+		needUpdate := true
+		for _, externalSubnet := range vpc.Status.ExternalSubnets {
+			if externalSubnet == subnet {
+				needUpdate = false
+				break
+			}
+		}
+		if needUpdate {
+			klog.Info("update externalSubnets status,add subnet : ", subnet)
```
Comment (bobz965) [inline pkg/controller/vpc.go:1174]: log 格式和输出有问题

File: pkg/controller/vpc.go [hunk 10]
```diff
@@ -1154,6 +1255,15 @@ func (c *Controller) handleDelVpcExternal(key string) error {
 	}
 	vpc := cachedVpc.DeepCopy()
 	vpc.Status.EnableExternal = cachedVpc.Spec.EnableExternal
+	if subnet != c.config.ExternalGatewaySwitch {
+		for i, externalSubnet := range vpc.Status.ExternalSubnets {
+			if externalSubnet == subnet {
+				klog.Info("update externalSubnets status,add subnet : ", subnet)
+				vpc.Status.ExternalSubnets = append(vpc.Status.ExternalSubnets[:i], vpc.Status.ExternalSubnets[i+1:]...)
```
Comment (bobz965) [inline pkg/controller/vpc.go:1262]: 感觉直接赋值就行，没必要一个个添加

**Review by bobz965 — CHANGES_REQUESTED** [review]

**Review by bobz965 — CHANGES_REQUESTED** [review]

File: pkg/controller/vpc.go [hunk 11]
```diff
@@ -564,12 +615,27 @@ func (c *Controller) handleAddOrUpdateVpc(key string) error {
...
-			// disconnect vpc to external
-			if err := c.handleDelVpcExternal(key); err != nil {
+			// disconnect vpc to default external
+			if err := c.handleDelVpcExternalSubnet(key, c.config.ExternalGatewaySwitch); err != nil {
 				klog.Errorf("failed to delete external connection for vpc %s, error %v", key, err)
 				return err
 			}
 		}
+
+		if !cachedVpc.Spec.EnableExternal && (cachedVpc.Status.AddExternalSubnets != nil || cachedVpc.Spec.AddExternalSubnets != nil) {
```
Comment (bobz965) [inline pkg/controller/vpc.go:625]:
```suggestion
		if cachedVpc.Status.AddExternalSubnets != nil && !cachedVpc.Spec.EnableExternal {
```

**Review by bobz965 — APPROVED** [review]

**Comment by bobz965** [comment]
hi @oilbeater ,  could you help review this ?

File: pkg/apis/kubeovn/v1/types.go [hunk 12]
```diff
@@ -415,12 +415,13 @@ type Vpc struct {
...
-	PolicyRoutes   []*PolicyRoute `json:"policyRoutes,omitempty"`
-	VpcPeerings    []*VpcPeering  `json:"vpcPeerings,omitempty"`
-	EnableExternal bool           `json:"enableExternal,omitempty"`
-	EnableBfd      bool           `json:"enableBfd,omitempty"`
+	Namespaces         []string       `json:"namespaces,omitempty"`
+	StaticRoutes       []*StaticRoute `json:"staticRoutes,omitempty"`
+	PolicyRoutes       []*PolicyRoute `json:"policyRoutes,omitempty"`
+	VpcPeerings        []*VpcPeering  `json:"vpcPeerings,omitempty"`
+	EnableExternal     bool           `json:"enableExternal,omitempty"`
+	AddExternalSubnets []string       `json:"addExternalSubnets,omitempty"`
```
Comment (oilbeater) [inline pkg/apis/kubeovn/v1/types.go:423]: We should name it `externalSubnets`

File: pkg/apis/kubeovn/v1/types.go [hunk 13]
```diff
@@ -481,6 +482,7 @@ type VpcStatus struct {
 	Subnets                 []string `json:"subnets"`
 	VpcPeerings             []string `json:"vpcPeerings"`
 	EnableExternal          bool     `json:"enableExternal"`
+	AddExternalSubnets      []string `json:"addExternalSubnets"`
```
Comment (oilbeater) [inline pkg/apis/kubeovn/v1/types.go:485]: No need to duplicate same filed in the status

**Review by oilbeater — CHANGES_REQUESTED** [review]

**Review by bobz965 — APPROVED** [review]

**Pull request merged**